src/gc.rs

```diff
@@ -0,0 +1,257 @@
+        this.ptr.as_ptr() as *const T
+    }
+
+    pub fn ptr_eq(this: &Self, other: &Self) -> bool {
+        this.ptr.as_ptr() == other.ptr.as_ptr()
+    }
+
+    pub fn from_raw(raw: *const T) -> Gc<T> {
+        Gc {
+            ptr: unsafe { NonNull::new_unchecked(raw as *mut GcBox<T>) },
+            _phantom: PhantomData,
+        }
+    }
+
+    fn from_inner(ptr: NonNull<GcBox<T>>) -> Self {
+        Self {
+            ptr,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T> Gc<MaybeUninit<T>> {
+    /// As with `MaybeUninit::assume_init`, it is up to the caller to guarantee
+    /// that the inner value really is in an initialized state. Calling this
+    /// when the content is not yet fully initialized causes immediate undefined
+    /// behavior.
+    pub unsafe fn assume_init(self) -> Gc<T> {
+        let ptr = self.ptr.as_ptr() as *mut GcBox<MaybeUninit<T>>;
+        Gc::from_inner((&mut *ptr).assume_init())
+    }
+}
+
+impl<T: ?Sized + fmt::Display> fmt::Display for Gc<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&**self, f)
+    }
+}
+
+/// A `GcBox` is a 0-cost wrapper which allows a single `Drop` implementation
+/// while also permitting multiple, copyable `Gc` references. The `drop` method
+/// on `GcBox` acts as a guard, preventing the destructors on its contents from
+/// running unless the object is really dead.
+struct GcBox<T: ?Sized>(ManuallyDrop<T>);
+
+impl<T> GcBox<T> {
+    fn new(value: T) -> *mut GcBox<T> {
+        let (layout, _) = Layout::new::<usize>().extend(Layout::new::<T>()).unwrap();
+        let base_ptr = unsafe { GC_ALLOCATOR.alloc(layout).unwrap().0.as_ptr() };
+        let gcbox = GcBox(ManuallyDrop::new(value));
+        let obj_ptr = unsafe { (base_ptr as *mut usize).add(1) } as *mut GcBox<T>;
+
+        unsafe {
+            obj_ptr.copy_from_nonoverlapping(&gcbox, 1);
+            let fatptr: &mut dyn Finalizable = &mut *obj_ptr;
+            let mut vptr = transmute::<*const dyn Finalizable, (usize, usize)>(fatptr).1;
+            ::std::ptr::write(
+                base_ptr as *mut Option<NonZeroUsize>,
+                Some(NonZeroUsize::new_unchecked(vptr)),
+            );
+            ::std::ptr::write(base_ptr as *mut Option<NonZeroUsize>, None);
+        }
+
+        forget(gcbox);
+        let flzr = Self::finalizer_shim;
+        unsafe {
+            boehm::GC_register_finalizer(
+                base_ptr as *mut _ as *mut ::std::ffi::c_void,
+                flzr,
+                ::std::ptr::null_mut(),
+                ::std::ptr::null_mut(),
+                ::std::ptr::null_mut(),
+            );
+        }
+        obj_ptr
+    }
+
+    unsafe extern "C" fn finalizer_shim(obj: *mut c_void, meta: *mut c_void) {
+        let vptr = *(obj as *mut Option<NonZeroUsize>);
+        match vptr {
+            Some(nzptr) => {
+                let objptr = (obj as *mut usize).add(1); // UB I think?
```

Comment:
Why would it be UB?

Reply:
It wouldn't. Sorry, stale comment.

Reply:
This is gone as part of e53a221a81efe70620a0126699cf3ff0e196ebd4